File: hot100/py/groupAnagrams.py
from typing import List


def groupAnagrams(strs: List[str]) -> List[List[str]]:
    """
    字母异位词分组

    使用字符频次数组作为 key，将异位词分组。
    时间复杂度: O(n*k)，其中 n 是字符串数量，k 是平均字符串长度
    空间复杂度: O(n*k)
    """
    char_map = {}

    for s in strs:
        # 统计字符频次
        freq = [0] * 26
        for char in s:
            freq[ord(char) - ord('a')] += 1

        # 用频次数组作为 key
        key = tuple(freq)

        if key not in char_map:
            char_map[key] = []
        char_map[key].append(s)

    return list(char_map.values())


if __name__ == "__main__":
    # 测试
    result = groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"])
    print(result)

File: hot100/js/groupAnagrams.js
/**
 * @description 字母异位词分组
 * @param {Array} strs 字符串数组
 * @returns {string[][]} 分组之后的异位词数组
 */
function groupAnagramsVersion1(strs) {
    const map = new Map();
    let baseNum = "a".charCodeAt();
    for (let str of strs) {
        let arr = new Array(26).fill(0);
        for (let s of str) {
            let n = s.charCodeAt() - baseNum
            arr[n] += 1;
        }
        let ns = arr.join(",")
        if (!map.has(ns)) {
            map.set(ns, [])
        }
        map.get(ns).push(str)
    }
    return Array.from(map.values())
}
console.log(groupAnagramsVersion1(["eat", "tea", "tan", "ate", "nat", "bat"]))